<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение." data-toc-modified-id="Введение.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение.</a></span></li><li><span><a href="#Обзор-данных." data-toc-modified-id="Обзор-данных.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных.</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры." data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></div>

# Исследование базы данных с помощью SQL.

## Введение.

Мы располагаем базой данных в которой хранятся информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. 

Описание данных

Таблица `books` - содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`- содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`- содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings` - содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`-содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

Цель исследования :

Выполнить задания с помощью SQL:

* Посчитайте, сколько книг вышло после 1 января 2000 года;

* Для каждой книги посчитайте количество обзоров и среднюю оценку;

* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


Ход исследования:

Провести обзор и предобработку данных

Сделать по одному SQL-запросу для решения каждого задания

Проанализировать результаты 

## Обзор данных.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [3]:
#функция для обзора данных
def overview_df(table):
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    data = pd.io.sql.read_sql(query, con = engine)
    
    print('Таблица', table )
    print('')
    print('...........Первые 5 строк...........')
    display(data.head())
    print('')
    print('')
    print('...........Тип данных...........')
    print('')
    print(data.info())
    print('')
    print('.................................................................................................................')

In [4]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    print(overview_df(table))

Таблица books

...........Первые 5 строк...........


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



...........Тип данных...........

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

.................................................................................................................
None
Таблица authors

...........Первые 5 строк...........


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



...........Тип данных...........

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

.................................................................................................................
None
Таблица publishers

...........Первые 5 строк...........


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company



...........Тип данных...........

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

.................................................................................................................
None
Таблица ratings

...........Первые 5 строк...........


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



...........Тип данных...........

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

.................................................................................................................
None
Таблица reviews

...........Первые 5 строк...........


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...



...........Тип данных...........

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None

.................................................................................................................
None


Данные представлены в корректном виде.

## Посчитайте, сколько книг вышло после 1 января 2000 года.

In [5]:
query1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-01-01'; 
''' 
request1 = pd.io.sql.read_sql(query1, con = engine)
request1

count
0    819

После 1 января 2000 года вышло **819 книг**.

## Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [6]:
query2 = ''' 
WITH a AS (SELECT r.book_id,
AVG(r.rating) AS avg_rating
FROM ratings AS r
GROUP BY r.book_id
),
b AS (SELECT book_id,
COUNT(review_id) AS reviews_cnt 
FROM reviews
GROUP BY book_id)

SELECT b.book_id,
b.reviews_cnt,
a.avg_rating
FROM b
FULL OUTER JOIN a ON a.book_id = b.book_id
''' 
request2 = pd.io.sql.read_sql(query2, con = engine)
request2

book_id  reviews_cnt  avg_rating
0        1.0          2.0    3.666667
1        2.0          1.0    2.500000
2        3.0          3.0    4.666667
3        4.0          2.0    4.500000
4        5.0          4.0    4.000000
..       ...          ...         ...
995    996.0          3.0    3.666667
996    997.0          3.0    3.400000
997    998.0          4.0    3.200000
998    999.0          2.0    4.500000
999   1000.0          4.0    3.833333

[1000 rows x 3 columns]

In [7]:
request2.describe()

book_id  reviews_cnt   avg_rating
count   994.000000   994.000000  1000.000000
mean    501.144869     2.809859     3.898973
std     288.845619     1.055873     0.562376
min       1.000000     1.000000     1.500000
25%     252.250000     2.000000     3.500000
50%     501.500000     3.000000     4.000000
75%     750.750000     3.000000     4.333333
max    1000.000000     7.000000     5.000000

In [8]:
query = ''' 
SELECT COUNT(*) 
FROM books
; 
''' 
request = pd.io.sql.read_sql(query, con = engine)
request

count
0   1000

Количество книг совпадает. 

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [9]:
query3 = ''' 
WITH a AS (SELECT book_id,
publisher_id
FROM books
WHERE num_pages > 50)

SELECT p.publisher,
COUNT(a.book_id) AS books_cnt
FROM a 
JOIN  publishers AS p ON a.publisher_id = p.publisher_id
GROUP BY p.publisher
ORDER BY books_cnt DESC
LIMIT 1
; 

''' 
request3 = pd.io.sql.read_sql(query3, con = engine)
request3

publisher  books_cnt
0  Penguin Books         42

Издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**. Это британское издательство, основанное в 1935 году в Лондоне. В 2001 году издательство стало проигрывать в США своему основному тамошнему конкуренту — издательству Random House. В 2013 году они объединились, создав первую в мире глобальную издательскую компанию. 

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [10]:
query4 = ''' 
WITH r AS (SELECT book_id,
COUNT(rating) AS rating_CNT,
AVG(rating) AS avg_rating
FROM ratings 
GROUP BY book_id
HAVING COUNT(rating) >= 50)

SELECT a.author,
AVG(r.avg_rating)
FROM authors AS a
JOIN books AS b ON a.author_id=b.author_id
RIGHT JOIN r ON r.book_id = b.book_id
GROUP BY a.author
ORDER BY AVG(r.avg_rating) DESC
LIMIT 1
; 

''' 
request4 = pd.io.sql.read_sql(query4, con = engine)
request4


author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

In [11]:
query40 = ''' 
SELECT a.author,
AVG(r.rating) AS avg_rating
FROM ratings AS r
JOIN books AS b ON b.book_id = r.book_id
JOIN authors AS a ON a.author_id = b.author_id
WHERE r.book_id IN (SELECT book_id
                    FROM ratings 
                    GROUP BY book_id
                    HAVING COUNT(book_id) >= 50)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1;

''' 
request40 = pd.io.sql.read_sql(query40, con = engine)
request40

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Автор с самой высокой средней оценкой книг(4,287) - **J.K. Rowling/Mary GrandPré** (среди авторов для книг с 50 и более оценками). Мари ГрандПре иллюстрировала все американские издания романов о Гарри Поттере автора Джоан Роулинг. 

Начиная с публикации первого романа «Гарри Поттер и философский камень» 26 июня 1997 года, книги серии обрели огромную популярность, признание критиков и коммерческий успех во всём мире. По состоянию на февраль 2018 года количество проданных книг составило около 500 миллионов экземпляров, вследствие чего серия вошла в список литературных бестселлеров. Романы переведены на 80 языков, в том числе на русский. Последние четыре книги последовательно поставили рекорды, как наиболее быстро продаваемые литературные произведения в истории. Поэтому неудивительно, что в нашей базе самая высокая средняя оценка среди книг с достаточно большим количеством оценок,  первое место принадлежит этой серии книг.

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
query5 = ''' 
WITH a AS (SELECT username,
COUNT(review_id) AS review_cnt
FROM reviews 
WHERE username IN (SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) > 50)
GROUP BY username) 

SELECT AVG(review_cnt)
FROM a 
''' 
request5 = pd.io.sql.read_sql(query5, con = engine)
request5

avg
0  24.333333

Пользователи, которые поставили больше 50 оценок в среднем писали 24 обзора.

## Вывод.

- Большинство книг базы вышло после 1 января 2000 года -819 из 1000.

- Количество обзоров на каждую книгу колеблется в диапазоне от 1 до 7. При среднем равном 3.

- Средняя оценка на каждую книгу достаточно высокая - при диапазоне средней оценки от 1,5 до 5 среднее составляет 4-ех.

- Издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**-  глобальная издательская компания, основанная около 90 лет назад. 

- Автор с самой высокой средней оценкой книг(4,287)- Джоан Роулинг, а именно серия книг о Гарри Поттере, иллюстрированная  Мари ГрандПре.

- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок составляет около 24 обзоров.